haskell/haddock.bzl

```diff
@@ -0,0 +1,165 @@
+
+  )
+  hoogle_file = ctx.actions.declare_file(
+    paths.replace_extension(ctx.rule.attr.name, ".txt"),
+    sibling = haddock_interface
+  )
+  input_sources = [ f for t in ctx.rule.attr.srcs for f in t.files.to_list() ]
+
+  args = ctx.actions.args()
+  args.add([
+    "-D", haddock_interface,
+    "--package-name={0}".format(ctx.rule.attr.name),
+    "--package-version={0}".format(ctx.rule.attr.version),
+    "-o", doc_dir,
+    "--html", "--hoogle",
+    # TODO: We shouldn't have to remember this, I think haskell
+    # targets should expose all GHC flags they were ultimately built
+    # with. That way we can pass exactly the same stuff. Having said
+    # that, it's not that easy: only some flags are relevant and only
+    # in some contexts.
+    "--optghc=-hide-all-packages",
+    "--title={0}".format(pkg_id),
+    # This is absolutely required otherwise GHC doesn't what package
+    # it's creating `Name`s for to put them in Haddock interface files
+    # which then results in Haddock not being able to find names for
+    # linking in environment after reading its interface file later.
+    "--optghc=-this-unit-id", "--optghc={0}".format(pkg_id)
+    # TODO: --hyperlinked-source or make a ticket
+  ])
+
+  dep_interfaces = depset()
+  for dep in ctx.rule.attr.deps:
+    if HaddockInfo in dep:
+      args.add("--read-interface={0},{1}".format(
+        # Is this the best we can do? We have to tell haddock where the
+        # docs for the given interface are and it has to be relative
+        # because bazel moves these things around but is relying on
+        # these always being in the same directory for the target the
+        # right thing to do? Feels too hacky.
+        paths.join("..", dep[HaddockInfo].doc_dir.basename),
+        dep[HaddockInfo].interface_file.path
+      ))
+      dep_interfaces = depset(transitive = [
+        dep_interfaces, depset([dep[HaddockInfo].interface_file])
+      ])
+
+  # Expose all prebuilt packages
+  for prebuilt_dep in ctx.rule.attr.prebuilt_dependencies:
+    args.add("--optghc=-package {0}".format(prebuilt_dep))
+
+  # Expose all bazel dependencies
+  for pkg_name in target[HaskellPackageInfo].names.to_list():
+    # Don't try to tell GHC to include the target itself.
+    if target[HaskellPackageInfo].name != pkg_name:
+      args.add("--optghc=-package {0}".format(pkg_name))
+
+  # Include all package DBs we know of. Technically we shouldn't
+  # include the one for target but it shouldn't hurt us.
+  for pkg_cache in target[HaskellPackageInfo].caches.to_list():
+    args.add("--optghc=-package-db {0}".format(pkg_cache.dirname))
+
+  # Pass same flags the user set for the build.
+  for ghc_flag in ctx.rule.attr.compiler_flags:
+    args.add("--optghc={0}".format(ghc_flag))
+
+  args.add(input_sources)
+
+  static_haddock_outputs = [
+    ctx.actions.declare_file(f, sibling=haddock_interface)
+    for f in ["doc-index.html", "haddock-util.js",
```

Comment:
Any chance we could avoid this hard coded list of Haddock artifacts?

Reply:
I don't really see how. We could list these out per-version after #56 but we need these one way or another.